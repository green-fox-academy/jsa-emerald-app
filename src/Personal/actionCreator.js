import { BACKEND_URL } from 'react-native-dotenv';

export const actionType = {
  LOGIN_START: 'LOGIN_START',
  LOGIN_FAILED: 'LOGIN_FAILED',
  LOGIN_SUCCESSFUL: 'LOGIN_SUCCESSFUL',
};

export function loginStart() {
  return {
    type: actionType.LOGIN_START,
  };
}

export function loginSuccessful(payload) {
  return {
    type: actionType.LOGIN_SUCCESSFUL,
    payload,
  };
}

export function loginFailed(payload) {
  return {
    type: actionType.LOGIN_FAILED,
    payload,
  };
}

export const requestLogin = (userInfo) => (dispatch) => {
  dispatch(loginStart());
  fetch(`${BACKEND_URL}/sessions`, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'Application/json',
    },
    body: JSON.stringify(userInfo),
  }).then((response) => response.json()).then((response) => {
    console.log(response);

    if (response.accessToken !== '' && response.accessToken !== undefined) {
      dispatch(loginSuccessful({
        email: userInfo.email,
        accessToken: response.accessToken,
      }));
      console.log(2);
    } else {
      dispatch(loginFailed({ message: response.message }));
      console.log(1);
      console.log(response.message);
    }
  }).catch((error) => {
    dispatch(loginFailed({ message: error }));
  });
};
