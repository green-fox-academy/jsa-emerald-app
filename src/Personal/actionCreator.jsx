import { BACKEND_URL } from 'react-native-dotenv';

export const actionType = {
  LOGIN_START: 'LOGIN_START',
  LOGIN_FAILED: 'LOGIN_FAILED',
  LOGIN_SUCCESSFUL: 'LOGIN_SUCCESSFUL',
};

export function loginStart() {
  return {
    type: actionType.LOGIN_START,
  };
}

export function loginSuccessful(userInfo) {
  return {
    type: actionType.LOGIN_SUCCESSFUL,
    payload: userInfo,
  };
}

export function loginFailed(userInfo) {
  return {
    type: actionType.LOGIN_FAILED,
    payload: userInfo,
  };
}

export const requestLogin = (userInfo) => (dispatch) => {
  dispatch(loginStart());
  fetch(`${BACKEND_URL}/sessions`, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'Application/json',
    },
    body: JSON.stringify(userInfo),
  }).then((response) => response.json()).then((response) => {
    if (response.accessToken !== '' && response.accessToken !== undefined) {
      dispatch(loginSuccessful({
        email: userInfo.email,
        accessToken: response.accessToken,
      }));
    } else {
      dispatch(loginFailed({ status: response.code, message: response.message }));
    }
  }).catch(() => {
    dispatch(loginFailed());
  });
};
