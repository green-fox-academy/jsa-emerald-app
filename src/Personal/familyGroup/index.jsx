import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { View, ScrollView } from 'react-native';
import setThemeStyle from '../../Common/theme/setThemeStyle';
import MainHeader from '../../Common/MainHeader';
import FamilyCard from './FamilyCard';
import SubmitBtn from '../SubmitBtn';
import AddMemberOverlay from './AddMemberOverlay';
import {
  cancelMemberSearch, updateFamilyMember, getFamilyMemberList, confirmFamilyMember,
} from './actionCreator';

export default function FamilyPage() {
  const { themeMode } = useSelector((state) => state.theme);
  const { members } = useSelector((state) => state.familyList);

  const styles = setThemeStyle(themeMode);
  const [isOverlayVisible, setOverlayVisibility] = useState(false);
  const dispatch = useDispatch();
  const addMemberName = (value) => {
    const searchIndex = members.findIndex((item) => item.id === value.id);
    if (searchIndex === -1) {
      dispatch(confirmFamilyMember(value));
      dispatch(updateFamilyMember());
      // dispatch(getFamilyMemberList());
    }
    setOverlayVisibility(false);
  };

  const cancelSearch = () => {
    dispatch(cancelMemberSearch());
    setOverlayVisibility(false);
  };

  useEffect(() => {
    console.log(members);

  }, []);

  useEffect(() => {
    dispatch(getFamilyMemberList());
  }, ['']);

  return (
    <View style={{ flex: 1, flexDirection: 'column' }}>
      <AddMemberOverlay
        isVisible={isOverlayVisible}
        onConfirm={(value) => addMemberName(value)}
        onCancel={() => cancelSearch()}
      />
      <MainHeader title="My Family" />
      <View style={[styles.deviceBody, { flex: 1, flexDirection: 'column' }]}>
        <View style={{ marginHorizontal: 10, marginTop: 15 }}>
          <SubmitBtn title="Add Member" disabled={false} onPressBtn={() => setOverlayVisibility(true)} />
        </View>
        <ScrollView>
          {members.map((value, idx) => <FamilyCard key={`ListGroup:${idx + 1}`} username={value.username} note="Father" imgUrl={value.url ? value.url : 'https://s3.amazonaws.com/uifaces/faces/twitter/ladylexy/128.jpg'} />)}
        </ScrollView>
      </View>
    </View>
  );
}
