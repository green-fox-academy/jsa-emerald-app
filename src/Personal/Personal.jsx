import React, { useState, useEffect } from 'react';
import {
  View, Text, ActivityIndicator, ScrollView, ImageBackground,
} from 'react-native';
import {
  Image, ListItem, Button,
} from 'react-native-elements';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import { useNavigation } from 'react-navigation-hooks';
import { requestBackup, requestRestore, logoutSuccessful } from './actionCreator';
import MainHeader from '../Common/MainHeader';
import RestoreOverlay from './RestoreOverlay';
import ThemeSelection from './themeSelection/ThemeSelection';
import setThemeStyle from '../Common/theme/setThemeStyle';
import profileImg from '../../assets/profile.jpg';

const avatar = require('../../assets/avatar.png');

export default function Personal() {
  const { navigate } = useNavigation();
  const dispatch = useDispatch();
  const backupState = useSelector((state) => state.backupState);
  const restoreState = useSelector((state) => state.restoreState);
  const user = useSelector((state) => state.user);
  const { themeMode } = useSelector((state) => state.theme);
  const styles = setThemeStyle(themeMode);

  const handleLogout = () => {
    // dispatch(requestBackup());
    dispatch(logoutSuccessful());
  };

  const [message, setMessage] = useState('Click to back up data');
  const [restoreMsg, setRestoreMsg] = useState('Click to restore data');
  const [isOverlayVisible, setOverlayVisibility] = useState(false);

  useEffect(() => {
    if (backupState.error !== '') {
      setMessage(backupState.error);
    }
    if (backupState.lastBackupDate !== null) {
      setMessage(`Last update: ${moment.unix(backupState.lastBackupDate).format('YYYY-MM-DD HH:mm')}`);
    }
    if (restoreState.lastRestoreDate !== null) {
      setRestoreMsg(`Last restore: ${moment.unix(restoreState.lastRestoreDate).format('YYYY-MM-DD HH:mm')}`);
    }
    if (restoreState.error !== '') {
      setRestoreMsg(restoreState.error);
    }
  }, [
    backupState.lastBackupDate,
    backupState.error,
    restoreState.lastRestoreDate,
    restoreState.error]);

  const restoreData = () => {
    dispatch(requestRestore());
    setOverlayVisibility(false);
  };

  return (
    <View style={{ flex: 1, flexDirection: 'column' }}>
      <MainHeader title="Profile" />
      <ScrollView style={[styles.deviceBody, { flex: 1, flexDirection: 'column' }]}>
        <RestoreOverlay
          isVisible={isOverlayVisible}
          onConfirm={() => restoreData()}
          onCancel={() => setOverlayVisibility(false)}
        />
        <View style={[styles.card, { padding: 0, overflow: 'hidden' }]}>
          <ImageBackground
            source={profileImg}
            style={{ borderTopLeftRadius: 6, borderTopRightRadius: 6 }}
          >
            <View style={{
              alignItems: 'center', paddingVertical: 30,
            }}
            >
              <Image source={avatar} style={{ width: 85, height: 85 }} />
              <Text style={[styles.secondaryHeading, { color: '#FAFAFA', fontWeight: 'bold', fontSize: 30 }]}>{user.username}</Text>
              <Text style={{ color: '#FAFAFA', fontWeight: 'bold', fontSize: 15 }}>{user.email}</Text>
            </View>
          </ImageBackground>
          <View style={{ paddingTop: 15, paddingBottom: 10, paddingHorizontal: 10 }}>
            <Text
              style={[styles.secondaryHeading, { marginLeft: 12, marginBottom: 5 }]}
            >
              General
            </Text>
            <ListItem
              title="Personal details"
              titleStyle={styles.listHeading}
              subtitle="you can edit your information about your email address, phone number"
              subtitleStyle={{ color: 'grey' }}
              bottomDivider
              containerStyle={{ backgroundColor: 'transparent' }}
            />
            <ListItem
              title="Backup data"
              titleStyle={styles.listHeading}
              subtitle={message}
              subtitleStyle={{ color: 'grey' }}
              rightElement={backupState.isInProgress ? <ActivityIndicator size="small" color="grey" />
                : (
                  <Button
                    title="Backup"
                    titleStyle={{ color: '#2fc899' }}
                    type="outline"
                    buttonStyle={{ borderColor: '#2fc899', borderRadius: 6 }}
                    onPress={() => dispatch(requestBackup())}
                  />
                )}
              containerStyle={{ backgroundColor: 'rgba(255, 255, 255, 0)' }}
              bottomDivider
            />
            <ListItem
              title="My family"
              titleStyle={styles.listHeading}
              containerStyle={{ backgroundColor: 'transparent' }}
              onPress={() => navigate('FamilyPage')}
              bottomDivider
            />
            <ListItem
              title="Change password"
              titleStyle={styles.listHeading}
              containerStyle={{ backgroundColor: 'rgba(255, 255, 255, 0)' }}
              bottomDivider
            />
            <ListItem
              title="Theme"
              titleStyle={styles.listHeading}
              rightElement={(
                <ThemeSelection />
              )}
              containerStyle={{ backgroundColor: 'rgba(255, 255, 255, 0)' }}
            />
          </View>
        </View>
        <View style={styles.card}>
          <Text style={[styles.secondaryHeading, { marginLeft: 12, marginBottom: 5 }]}>
            Cloud Service
          </Text>
          <ListItem
            title="Restore data"
            titleStyle={styles.listHeading}
            subtitle={restoreMsg}
            subtitleStyle={{ color: 'grey' }}
            rightElement={restoreState.isInProgress ? <ActivityIndicator size="small" color="grey" />
              : (
                <Button
                  title="Restore"
                  titleStyle={{ color: '#2fc899' }}
                  type="outline"
                  buttonStyle={{ borderColor: '#2fc899', borderRadius: 6 }}
                  onPress={() => setOverlayVisibility(true)}
                />
              )}
            bottomDivider
            containerStyle={{ backgroundColor: 'rgba(255, 255, 255, 0)' }}
          />
          <ListItem
            title="Banking Sync"
            titleStyle={styles.listHeading}
            subtitle="Sync your transactions from bank"
            subtitleStyle={{ color: 'grey' }}
            onPress={() => navigate('OpenBanking')}
            containerStyle={{ backgroundColor: 'rgba(255, 255, 255, 0)' }}
          />
          <Button title="Logout" color="#f194ff" onPress={handleLogout} />
        </View>
      </ScrollView>
    </View>
  );
}
