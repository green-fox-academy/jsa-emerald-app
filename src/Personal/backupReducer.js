import { actionType } from './actionCreator';

const initialBackup = {
  isInProgress: false,
  error: '',
  lastBackupDate: null,
};

const backupState = (state = initialBackup, action) => {
  switch (action.type) {
    case actionType.BACKUP_START:
      return {
        ...state,
        isInProgress: true,
      };
    case actionType.BACKUP_SUCCESSFUL:
      return {
        ...state,
        isInProgress: false,
        lastBackupDate: action.payload,
      };
    case actionType.BACKUP_FAILED:
      return {
        ...state,
        isInProgress: false,
        lastBackupDate: 0,
      };
    default:
      return state;
  }
};

export default backupState;
