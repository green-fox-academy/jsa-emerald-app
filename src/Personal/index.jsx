import React, { useState, useEffect } from 'react';
import { View, Text, Alert } from 'react-native';
import { Avatar, ListItem } from 'react-native-elements';
import { useDispatch, useSelector } from 'react-redux';
import { requestBackup } from './actionCreator';
import MainHeader from '../Common/MainHeader';
import styles from '../Common/themeStyle';

export default function Personal() {
  const [switchValue, setSwitchValue] = useState(false);
  const dispatch = useDispatch();
  const { transactions } = useSelector((state) => state.transactions);
  const backupState = useSelector((state) => state.backupState);

  useEffect(() => {
    if (switchValue) {
      dispatch(requestBackup(transactions));
      if (backupState.stateCode === 200) {
        Alert.alert('Backup successfully!');
      } else {
        Alert.alert("There's something wrong, please try again later");
      }
    }
  }, [switchValue]);

  return (
    <View style={{ flex: 1, flexDirection: 'column' }}>
      <MainHeader title="Profile" />
      <View style={[styles.deviceBody, { flex: 1, flexDirection: 'column' }]}>
        <View style={[styles.card, { padding: 0 }]}>
          <View style={{
            alignItems: 'center', backgroundColor: '#f078a4', borderTopLeftRadius: 6, borderTopRightRadius: 6, paddingVertical: 30,
          }}
          >
            <Avatar rounded size={60} icon={{ name: 'user', type: 'font-awesome' }} />
            <Text style={[styles.secondaryHeading, { color: 'white' }]}>Johnathan Doe</Text>
            <Text style={{ color: 'white' }}>Email</Text>
          </View>
          <View style={{ paddingTop: 15, paddingBottom: 10, paddingHorizontal: 10 }}>
            <Text style={[styles.secondaryHeading, { color: 'grey', marginLeft: 12, marginBottom: 5 }]}>General</Text>
            <ListItem
              title="Personal details"
              subtitle="you can edit your information about your email address, phone number"
              subtitleStyle={{ color: 'grey' }}
              bottomDivider
            />
            <ListItem
              title="Backup data"
              subtitle="Login to backup data"
              subtitleStyle={{ color: 'grey' }}
              switch={{
                disabled: false, trackColor: { true: '#5C6BC0' }, value: switchValue, onValueChange: () => setSwitchValue(!switchValue),
              }}
              bottomDivider
            />
            <ListItem
              title="Change Password"
              bottomDivider
            />
            <ListItem title="Categories" subtitle="Add or remove categories" subtitleStyle={{ color: 'grey' }} />
          </View>
        </View>
      </View>
    </View>
  );
}
