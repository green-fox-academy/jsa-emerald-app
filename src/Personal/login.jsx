import React, { useState, useEffect } from 'react';
import {
  View, Text, KeyboardAvoidingView, Platform,
} from 'react-native';
import { Input, Icon } from 'react-native-elements';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';
import { requestRestore, requestLogin } from './actionCreator';
import MainHeader from '../Common/MainHeader';
import utils from './utils';
import LoginView from './loginView';

import SubmitBtn from './submitBtn';

export default function Login() {
  const { navigate } = useNavigation();
  const user = useSelector((state) => state.user);
  const dispatch = useDispatch();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [passwordHidden, setPasswordHidden] = useState(true);
  const [errorMsgDisplay, setErrorMsgDisplay] = useState(false);

  const login = (userInfo) => {
    dispatch(requestLogin(userInfo));
  };

  useEffect(() => {
    if (user.accessToken) navigate('Index');
    // dispatch(requestRestore());
  }, [user.accessToken]);

  useEffect(() => {
    if (user.status !== '' && user.status !== undefined) {
      setErrorMsgDisplay(true);
    }
  }, [user.status]);

  return (
    <>
      <View style={{ borderBottomWidth: 2, borderBottomColor: '#f8f8f8' }}>
        <MainHeader title="Login" />
      </View>
      <KeyboardAvoidingView
        style={LoginView.registerForm}
        behavior="padding"
        keyboardVerticalOffset={Platform.OS === 'ios' ? 64 : 20}
        enabled
      >
        <View>
          <View style={LoginView.inputSection}>
            <Input
              placeholder="email@address.com"
              leftIcon={(
                <Icon
                  name="email"
                  color="#a8a8a8"
                />
          )}
              leftIconContainerStyle={LoginView.inputIcon}
              label="Your Email Address"
              textContentType="emailAddress"
              value={email}
              onChangeText={(value) => setEmail(value)}
              labelStyle={{ color: 'gray' }}
            />
            <Text style={LoginView.note}>
              {utils.emailValidation(email) || email === '' ? '' : 'Your email is not correct.' }
            </Text>
          </View>
          <View style={LoginView.inputSection}>
            <Input
              placeholder="Password"
              leftIcon={(
                <Icon
                  name="lock"
                  color="#a8a8a8"
                />
              )}
              leftIconContainerStyle={LoginView.inputIcon}
              label="Your password"
              secureTextEntry={passwordHidden}
              value={password}
              onChangeText={(value) => setPassword(value)}
              labelStyle={{ color: 'gray' }}
              rightIcon={(
                <Icon
                  name={`md-eye${passwordHidden ? '-off' : ''}`}
                  type="ionicon"
                  color="gray"
                  onPress={() => setPasswordHidden(!passwordHidden)}
                />
            )}
            />
          </View>
          {
             (errorMsgDisplay) ? (
               <View style={[LoginView.inputSection, LoginView.errorBox]}>
                 <Text style={LoginView.errorText}>{user.message}</Text>
               </View>
             ) : (null)
          }
        </View>
        <View>
          <SubmitBtn
            disabled={utils.loginValidation(email, password) === false}
            onPressBtn={() => login({ email, password })}
          />
        </View>
      </KeyboardAvoidingView>
    </>
  );
}
