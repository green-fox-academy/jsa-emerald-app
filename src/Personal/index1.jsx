import React from 'react';
import { View, Button } from 'react-native';
import { useNavigation } from 'react-navigation-hooks';
import { useSelector } from 'react-redux';
import styles from './loginView';
import Personal from './personal';

export default function Index() {
  const { navigate } = useNavigation();
  const user = useSelector((state) => state.user);

  return (
    (user.accessToken !== '' && user.accessToken !== undefined) ? <Personal /> : (
      <View style={styles.container}>
        <Button title="Login" onPress={() => navigate('Login')} />
      </View>
    )
  );
}
