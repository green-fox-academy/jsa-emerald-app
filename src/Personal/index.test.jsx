import React from 'react';
import renderer from 'react-test-renderer';
import Personal from './index';

describe('Personal', () => {
  it('render ', () => {
    const tree = renderer.create(<Personal />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
