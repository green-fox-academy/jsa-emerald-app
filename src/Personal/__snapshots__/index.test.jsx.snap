// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Personal render  1`] = `
<View>
  <Text>
    Personal Screen
  </Text>
</View>
`;
